src/main/java/com/powsybl/openloadflow/ac/ReactiveLimitsOuterLoop.java

```diff
@@ -53,27 +51,28 @@ public String getType() {
 
     public static void switchPvPq(LfBus bus, EquationSystem equationSystem, VariableSet variableSet, double newGenerationTargetQ) {
         bus.setGenerationTargetQ(newGenerationTargetQ);
-        bus.setVoltageControl(false);
+        bus.setVoltageControlEnabled(false);
 
         Equation qEq = equationSystem.createEquation(bus.getNum(), EquationType.BUS_Q);
         qEq.setActive(true);
 
-        LfBus controlledBus = bus.getControlledBus().orElse(null);
-        if (controlledBus != null) {
-            updateControlledBus(controlledBus, equationSystem, variableSet);
-        } else {
-            Equation vEq = equationSystem.createEquation(bus.getNum(), EquationType.BUS_V);
-            vEq.setActive(false);
-        }
+        bus.getVoltageControl().ifPresentOrElse(
+            voltageControl -> updateControlledBus(voltageControl, equationSystem, variableSet),
+            () -> equationSystem.createEquation(bus.getNum(), EquationType.BUS_V).setActive(false)
+        );
     }
 
-    public static void updateControlledBus(LfBus controlledBus, EquationSystem equationSystem, VariableSet variableSet) {
+    public static void updateControlledBus(VoltageControl voltageControl, EquationSystem equationSystem, VariableSet variableSet) {
+
+        LfBus controlledBus = voltageControl.getControlledBus();
+        Set<LfBus> controllerBuses = voltageControl.getControllerBuses();
+
         // clean reactive power distribution equations
-        controlledBus.getControllerBuses().forEach(b -> equationSystem.removeEquation(b.getNum(), EquationType.ZERO_Q));
+        controllerBuses.forEach(b -> equationSystem.removeEquation(b.getNum(), EquationType.ZERO_Q));
 
         // controlled bus has a voltage equation only if one of the controller bus has voltage control on
-        List<LfBus> controllerBusesWithVoltageControlOn = controlledBus.getControllerBuses().stream()
-                .filter(LfBus::hasVoltageControl)
+        List<LfBus> controllerBusesWithVoltageControlOn = controllerBuses.stream()
+                .filter(LfBus::isVoltageController)
```

Comment:
This is weird even if I understand why.

Reply:
Maybe we should add for clarity some kind of `LfBus::isVoltageControllerEnabled()` method? or a `VoltageControl::getEnabledControllerBuses()`?